### PR TITLE
extract.cpp: Fixed handling multi-output files, added some debug

### PR DESCRIPTION
**Pull request by Michal-Ston-Mobica**
Changed logic in handling multi-output files to write them in sequence, not chunk-by-chunk.
This should fix some errors in writing output ZIPs and general file handling for some installers.

### CODE REVIEW

File: src/wasm/extract.hpp [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 #include "wasm/fdzipstream/fdzipstream.h"
 
+#define wasm_debug(X) do { if (extractor::get().get_settings().debug_messages_enabled) log_info << X; } while(0)
```
Comment (mrcebryk) [inline src/wasm/extract.hpp]: Do we need that construction here?
`do { ... } while(0)`

File: src/wasm/extract.cpp [hunk 2]
```diff
@@ -917,28 +930,61 @@ std::string extractor::extract(const std::string& list_json) {
...
+              // Reload chunk containing the duplicated file and its offset
+              chunk_source = stream::chunk_reader::get(*slice_reader, chunk.first, "");
+              offset = offset_start;
+            }
+          }
 
-          json ret;
-          ret["status"] = "Aborted by user";
-          return ret.dump();
+          wasm_debug("pre-copy chunk=" << chunk.first.first_slice << " " << chunk.first.last_slice
```
Comment (mrcebryk) [inline src/wasm/extract.cpp]: I would rather consider using different format of logging function like:
`wasm_log << "pre-copy chunk=" ....`
or
`conditional_log(extractor::get().get_settings().debug_messages_enabled) << "pre-copy chunk=" ....`

Would be more readable in my opinion.

File: src/wasm/extract.hpp (hunk 1 above)
Comment (mrcebryk) [inline src/wasm/extract.hpp]: Since you use inside log_info I would recommend to change this macro name to: wasm_info

File: src/wasm/extract.cpp [hunk 3]
```diff
@@ -64,8 +70,12 @@ bool file_output::write(char* data, size_t n) {
 
   position_ += n;
   total_written_ += n;
-
-  if (is_complete()) {
+  if (extractor::get().get_settings().debug_messages_enabled) {
```
Comment (mrcebryk) [inline src/wasm/extract.cpp:68]: It would be good to unify logging handling, looks like you could use the "wasm_debug" here.

File: src/wasm/extract.cpp [hunk 4]
```diff
@@ -842,19 +855,19 @@ uint64_t extractor::seek_input_stream(stream::chunk_reader::type* chunk_source,
...
+    wasm_debug("copy_data read, extracted_n=" << extracted_n);
+
     if (extracted_n > 0 || (source->eof() && !source->bad())) {
-      for (auto output : outputs) {
-        bool success = output->write(buffer, extracted_n);
-        if (!success) {
-          throw std::runtime_error("Error writing file \"" + output->path().string() + '"');
-        }
+      bool success = output->write(buffer, extracted_n, file.size);
+      if (!success) {
```
Comment (mrcebryk) [inline src/wasm/extract.cpp]: Since you don't use that "success" variable anywhere below, I would recommend to simplify that code to:
     ` if (!output->write(buffer, extracted_n, file.size)) {`

File: src/wasm/extract.cpp [hunk 5]
```diff
@@ -64,8 +70,12 @@ bool file_output::write(char* data, size_t n) {
...
   position_ += n;
   total_written_ += n;
-
-  if (is_complete()) {
+  if (extractor::get().get_settings().debug_messages_enabled) {
+    printf("write: position=%llu, total_written=%llu, fsize=%llu, size2=%zu, n=%zu\n", position_,
+           total_written_, file_->entry().size, size2, n);
+    std::cout << "write: is_complete():" << std::endl;
+  }
+  if (file_->entry().size != 0 && is_complete() || size2 == total_written_) {
```
Comment (mrcebryk) [inline src/wasm/extract.cpp:73]: We will have different combinations of situations when we don't call:
`zs_entryend(zip_, zip_entry_, 0);`

I assume some of them might be expected, others not, I would recommend to add proper logs, especially warnings/errors in unexpected cases.

**Review by mrcebryk — COMMENTED** [review]
Interesting changes, I would consider improving the format of introduced wasm_debug macro and to add additional logs (or exceptions?) in unexpected cases

File: src/wasm/extract.cpp (hunk 3 above)
Comment (Michal-Ston-Mobica) [reply src/wasm/extract.cpp:68]: This was done on purpose to potentially save some execution time, as there are two consecutive log writes handled by a single _if_ statement, which would make two _ifs_ when used with a _wasm_debug_.
It's not obvious whether a simple _if_ like this is optimized when compiling to WASM, and the main goal of this commit was to expand on actual extract functionality while keeping some of the added debug as an extra and not to overhaul the whole debug system.
The existing debug calls in _extract.cpp_ weren't even operational before introducing _wasm_debug_, for an unknown reason.

File: src/wasm/extract.cpp (hunk 5 above)
Comment (Michal-Ston-Mobica) [reply src/wasm/extract.cpp:73]: The original is_complete() only checks whether file_->entry().size == total_written_, while this parameter is not always set for an unknown, probably installer-related reason. Thus an additional parameter of size2 with the actual file size that comes from another file descriptor was added to fix this issue and ensure the file is written properly.
This might've not been an issue with the original code, as it was only writing directly to the filesystem, and closing the files was always done at least by the OS when exiting the process, or somehow done via C++ interface.

To sum up: this whole _if_ statement performs a single check - to see if the file was written completely. Maybe it could be rewritten to be a bit more clear, but it's not simply a matter of rewriting is_complete(), as it's not aware of the size2, and it's not a part of the object it's being called on.

File: src/wasm/extract.cpp (hunk 4 above)
Comment (Michal-Ston-Mobica) [reply src/wasm/extract.cpp]: Valid point, although you can probably see this part was already present in the code, probably to make this part a bit more readable.
I agree that it can and probably should be simplified at one point or another, although a general refactor wasn't my goal when making this PR.

File: src/wasm/extract.cpp (hunk 2 above)
Comment (Michal-Ston-Mobica) [reply src/wasm/extract.cpp]: This function (or more specifically macro) was meant to be used only for debug and it's only used in wasm-related code, so I don't really see the point of making this more general or rewriting to be used like a stream.
It might be a good idea to make it more readable, but I assumed it's good enough for now and hoped that it could be addressed later, as important issues should resolved first and without much delay.

File: src/wasm/extract.hpp (hunk 1 above)
Comment (Michal-Ston-Mobica) [reply src/wasm/extract.hpp]: It's a popular way to create a macro wrapper that doesn't produce any errors, no matter how it's used in the code.
wasm_info is just wrapped here with an additional conditional statement, making this output the log only if the debug output is enabled before starting the extraction.
